### DISCUSSION TITLE
How to write css output before/after each plugin?

### DISCUSSION THREAD

**Discussion by therealgilles** [postcss/postcss#1573] [Q&A]
I am trying to track down an incorrect css build and would like to figure out which plugin is at fault. Is there a way to write out the css output before/after each plugin?

**Comment by ai** [comment]
There is no easy way, because in PostCSS 8 all plugins runs together. PostCSS walks through CSS tree and send nodes to listeners from all plugins.

You can try to debug code in `walkSync` and `async` of https://github.com/postcss/postcss/blob/main/lib/lazy-result.js

**Reply by therealgilles** [reply]
Thanks for your response. It would be useful to have a mode where each plugin is run separately. It would make it easier to isolate issues.

**Reply by ai** [reply]
> It would be useful to have a mode where each plugin is run separately

It will change how plugin intract between each other and potentially hide the problem.

**Reply by therealgilles** [reply]
Question: how do plugins figure out which what needs to run first?
For instance, if I use postcss-nested to un-nest some CSS code before it can be processed by another plugin (that does not supported nested css), how does that work?

**Reply by ai** [reply]
PostCSS 8 plugins are not “simple functions” anymore.

They are list of events https://github.com/postcss/postcss/blob/main/docs/writing-a-plugin.md#step-3-find-nodes

When you add plugin to PostCSS you add plugin’s evens to global events list.

Then PostCSS walk through CSS AST nodes. For each node it looks for event and run them.

PostCSS 7 plugin runs before this global walk as “simple functions” one-by-one.

**Reply by therealgilles** [reply]
Okay, so how are events between plugins ordered?

**Reply by ai** [reply]
By the order in PostCSS config

**Reply by therealgilles** [reply]
I'm trying to debug an issue related to tailwindcss and postcss-nested during tailwindcss production build. Some properties get incorrectly merged with their parent properties. If I disable tailwindcss, the incorrect merging does not happen. My current thinking is that somehow tailwindcss runs before postcss-nested and as it does not properly support nesting, it messes things up. Any idea how could go about debugging and/or fixing this issue?

**Reply by ai** [reply]
Try to debug methods which change AST (in `Node` and `Container`) to see when somebody changes AST.

**Reply by therealgilles** [reply]
Is there a way to list each event and what plugin it belongs to?

**Reply by ai** [reply]
Add `console.log` to this method https://github.com/postcss/postcss/blob/main/lib/lazy-result.js#L402

**Reply by therealgilles** [reply]
When I do that, tailwindcss does not appear, I only see the following plugins:

  postcssPlugin: 'postcss-nested'
  postcssPlugin: 'postcss-reporter'
  postcssPlugin: 'autoprefixer'
  postcssPlugin: 'postcss-import-parser'
  postcssPlugin: 'postcss-url-parser'

**Reply by therealgilles** [reply]
Also missing is postcss-preset-env.

**Reply by therealgilles** [reply]
Oh maybe I did not put the console.log statement in the right place.

**Reply by ai** [reply]
Or these plugins use PostCSS 7 model

**Reply by therealgilles** [reply]
How do PostCSS 7 vs PostCSS 8 plugins interact?
I do see some listed as objects and others as functions (even some anonymous functions).

**Reply by ai** [reply]
PostCSS  runs PostCSS 7 one by one. Then it starts walk through CSS with all PostCSS 8 plugins together.

**Reply by therealgilles** [reply]
Hmm tailwindcss is advertised to be using PostCSS 8 and I can see its object declaration:

```JS
  return {
    postcssPlugin: 'tailwindcss',
    plugins: [...plugins, (0, _processTailwindFeatures.default)(getConfig), _formatCSS.default]
  };
```

so not sure why it does not appear in my console.log.

**Reply by ai** [reply]
`Once` event runs together with PostCSS 7 plugins

**Reply by therealgilles** [reply]
Is there a way to put a PostCSS 8 plugin in the PostCSS 7 plugin queue so that it runs on its own in a controlled order?

**Reply by therealgilles** [reply]
I am also looking at ways to having completely separated PostCSS runs.

**Reply by ai** [reply]
`Once` from PostCSS 8 plugin works in PostCSS 7 plugins queue.

There is no way to use full event system from PostCSS 8 to PostCSS 7 queue.

Maybe you can show CSS before/after very event? Or you can track AST changes by hacking `Node.*` methods.
